aptos-move/framework/aptos-framework/sources/delegation_pool.move

```diff
@@ -0,0 +1,1719 @@
+    /// For latter case, extracted coins will be used to buy shares into the active pool and
+    /// escape inactivation when current lockup ends.
+    fun redeem_inactive_shares(
+        pool: &mut DelegationPool,
+        shareholder: address,
+        coins_amount: u64,
+        lockup_cycle: u64,
+    ): u64 {
+        let inactive_shares = table::borrow_mut(&mut pool.inactive_shares, lockup_cycle);
+        let shares_to_redeem = amount_to_shares_to_redeem(inactive_shares, shareholder, coins_amount);
+        if (shares_to_redeem == 0) return 0;
+        let redeemed_coins = pool_u64::redeem_shares(inactive_shares, shareholder, shares_to_redeem);
+
+        // if delegator entirely reactivated pending_inactive stake or withdrawn inactive stake,
+        // enable unlocking again by deleting this pending withdrawal
+        if (pool_u64::shares(inactive_shares, shareholder) == 0) {
+            table::remove(&mut pool.pending_withdrawals, shareholder);
+        };
+        // destroy inactive-shares pool of past OLC if all its stake has been withdrawn
+        if (lockup_cycle < pool.observed_lockup_cycle && total_coins(inactive_shares) == 0) {
+            pool_u64::destroy_empty(table::remove<u64, pool_u64::Pool>(&mut pool.inactive_shares, lockup_cycle));
+        };
+
+        redeemed_coins
+    }
+
+    /// Calculate stake deviations between the delegation and stake pools in order to
+    /// capture the rewards earned in the meantime, resulting operator commission and
+    /// whether the lockup expired on the stake pool.
+    fun calculate_stake_pool_drift(pool: &DelegationPool): (bool, u64, u64, u64, u64) {
+        let (active, inactive, pending_active, pending_inactive) = stake::get_stake(get_pool_address(pool));
+        assert!(
+            inactive >= pool.total_coins_inactive,
+            error::invalid_state(ESLASHED_INACTIVE_STAKE_ON_PAST_OLC)
+        );
+        // determine whether a new lockup cycle has been ended on the stake pool and
+        // inactivated SOME `pending_inactive` stake which should stop earning rewards now,
+        // thus requiring separation of the `pending_inactive` stake on current observed lockup
+        // and the future one on the newly started lockup
+        let lockup_cycle_ended = inactive > pool.total_coins_inactive;
+
+        // actual coins on stake pool belonging to the active shares pool
+        active = active + pending_active;
+        // actual coins on stake pool belonging to the shares pool hosting `pending_inactive` stake
+        // at current observed lockup cycle, either pending: `pending_inactive` or already inactivated:
+        if (lockup_cycle_ended) {
+            // `inactive` on stake pool = any previous `inactive` stake +
+            // any previous `pending_inactive` stake and its rewards (both inactivated)
+            pending_inactive = inactive - pool.total_coins_inactive
```

Comment:
This is the new inactive stake portion where commission has not been extracted from yet right?

Reply:
yes, if stake has been inactivated in the meantime, should still detect what was the end amount of pending_inactive stake when the lockup expired, specifically pending_inactive rewards earned since last sync and until stake has been inactivated